implement hasValue method

ItIsGreg opened the pull request:
The hasValue() method of the Fhirpath specification was not implemented (see also [issue ](https://github.com/HL7/fhirpath.js/issues/120). 

I added hasValue to the invocation table and provided an implementation in misc.js. The function basically checks if the content of the collection to be evaluated is a object or not.

The hasValue() method is integral to the constraints specified in StructureDefinitions. Without an implementation the evaluation of constraints with fhirpath.js is not possible.

plynchnlm commented:
Thanks for the PR.  I am not sure at the moment whether checking if the value is object or not is sufficient, but it might be.  Could you add a test?

ItIsGreg commented:
Thanks for the review! I added 3 test cases:

- the input collections contains a string
- the input collections contains multiple strings
- the input collection contains an object

I modified the hasValue implementation to check for input collection length as I believe, according to the [specification](https://hl7.org/fhir/fhirpath.html) it should only return true if a single value is given.

Please feel free to suggest further tests!

yuriy-sedinkin commented on src/misc.js:
```diff
@@ -265,4 +265,19 @@ engine.singleton = function (coll, type) {
+ */
+engine.hasValueFn = function(coll) {
+  if (coll.length === 1){
+    return [isPrimitive(coll[0].data)];
```
We are currently using `util.valData(coll[0])` instead of `coll[0].data`, but I guess it is not a big deal in this case.
Maybe I'm misunderstanding the spec, but it says "Otherwise, the return value is empty."
See http://hl7.org/fhir/fhirpath.html#functions
In JS implementation, empty is an empty array.

ItIsGreg replied:
Concerning the specification, I concur with you regarding the empty array. Strangely the [simplifier implementation](https://simplifier.net/fhirpath?filepath=package/CapabilityStatement-base.json&scope=hl7.fhir.r4.core@4.0.1&query=) and the [pascal reference implementation](https://github.com/HealthIntersections/fhirserver/blob/master/library/fhir4/fhir4_pathengine.pas) both return false... which made me stumble a bit... But yeah, I guess according to the implementation an empty array should be returned.

plynchnlm replied:
Grahame says it can return false, so there may be three possible return values.  https://chat.fhir.org/#narrow/stream/179266-fhirpath/topic/hasValue

yuriy-sedinkin commented on src/misc.js:
```diff
@@ -265,4 +265,19 @@ engine.singleton = function (coll, type) {
+};
+
+function isPrimitive(data){
+  return data !== null && typeof data !== "object";
```
`typeof null === 'object'` so we don't need the first check.

@plynchnlm I'm wondering if we also need to check the node type in the model.
This will affect performance. What do you think?
We can read the node type from `coll[0].path` when a model is specified.
The model can be specified by adding: `model: r4` to the test description in hasValue.yaml
The list of primitive types is here:
https://hl7.org/fhir/datatypes.html#primitive

yuriy-sedinkin replied:
@plynchnlm If no model is specified, should this method work?

plynchnlm replied:
I think if no FHIR model is specified, hasValue need not work, since it is a FHIR extension.  However, I suppose you could have two implementations, one slower one checking type when there isn't a model, and one faster one when the model is available.

ItIsGreg commented:
Thanks again for the suggestions. I have now implemented node type checking against a list of primitives derived from [here](https://hl7.org/fhir/datatypes.html#primitive) if a model is specified and a default checking against javascript primitive types.
Furthermore I added test cases for different models.

Please let me know what you think about the changes made. Looking forward to your reviews!

ItIsGreg commented on src/misc.js:
```diff
@@ -268,16 +268,54 @@ engine.singleton = function (coll, type) {
+    case 'string':
+    case 'number':
+    case 'boolean':
+    case 'symbol':
```
Thanks a lot for pointing out this question mark. I derived the list from the javascript primitive types. Symbol was also a type where I was not sure whether there will be any usage of. Thanks for clarifying. I removed the line accordingly.

ItIsGreg commented on src/misc.js:
```diff
@@ -268,16 +268,54 @@ engine.singleton = function (coll, type) {
+    case 'number':
+    case 'boolean':
+    case 'symbol':
+    case 'bigint':
```
Thank you for the insight into which type are currently used in fhirpath.js. As BigInt is a primitve type in the FHIR-specification, I included it in the list for default type checking. For now, I commented it out as suggested however!

plynchnlm commented:
This looks good to us now.  We'll get it merged and published.

ItIsGreg commented:
Thanks for the high quality feedback during the process (:
